libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -0,0 +1,332 @@
+    libsnark::Fqe_mul_gadget<ppT> check_out_Rz;
+
+    // ell_0 = xi * I
+    // ell_vw = -H
+    // ell_vv = 3 * J
+    bls12_377_ate_ell_coeffs<ppT> out_coeffs;
+
+    bls12_377_ate_dbl_gadget(
+        libsnark::protoboard<FqT> &pb,
+        const bls12_377_G2_proj<ppT> &R,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+
+    // R should already be assigned. Computes all internal values and
+    // outResult.
+    void generate_r1cs_witness(const libff::Fr<ppT> &two_inv);
+};
+
+template<typename ppT>
+class bls12_377_ate_add_gadget : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    typedef libff::Fq<libsnark::other_curve<ppT>> FqT;
+    typedef libff::Fqe<libsnark::other_curve<ppT>> FqeT;
+
+    libsnark::Fqe_variable<ppT> base_X;
+    libsnark::Fqe_variable<ppT> base_Y;
+    bls12_377_G2_proj<ppT> in_R;
+
+    // A = Qy * Rz;
+    libsnark::Fqe_variable<ppT> A;
+    libsnark::Fqe_mul_gadget<ppT> check_A;
+    // B = Qx * Rz;
+    libsnark::Fqe_variable<ppT> B;
+    libsnark::Fqe_mul_gadget<ppT> check_B;
+    // theta = Ry - A;
+    libsnark::Fqe_variable<ppT> theta;
+    // lambda = Rx - B;
+    libsnark::Fqe_variable<ppT> lambda;
+    // C = theta.squared();
+    libsnark::Fqe_variable<ppT> C;
+    libsnark::Fqe_sqr_gadget<ppT> check_C;
+    // D = lambda.squared();
+    libsnark::Fqe_variable<ppT> D;
+    libsnark::Fqe_sqr_gadget<ppT> check_D;
+    // E = lambda * D;
+    libsnark::Fqe_variable<ppT> E;
+    libsnark::Fqe_mul_gadget<ppT> check_E;
+    // F = Rz * C;
+    libsnark::Fqe_variable<ppT> F;
+    libsnark::Fqe_mul_gadget<ppT> check_F;
+    // G = Rx * D;
+    libsnark::Fqe_variable<ppT> G;
+    libsnark::Fqe_mul_gadget<ppT> check_G;
+    // H = E + F - (G + G);
+    libsnark::Fqe_variable<ppT> H;
+    // I = Ry * E;
+    libsnark::Fqe_variable<ppT> I;
+    libsnark::Fqe_mul_gadget<ppT> check_I;
+    // J = theta * Qx - lambda * Qy;
+    libsnark::Fqe_variable<ppT> theta_times_Qx;
+    libsnark::Fqe_mul_gadget<ppT> check_theta_times_Qx;
+    libsnark::Fqe_variable<ppT> lambda_times_Qy;
+    libsnark::Fqe_mul_gadget<ppT> check_lambda_times_Qy;
+    libsnark::Fqe_variable<ppT> J;
+
+    // out_Rx = lambda * H;
+    libsnark::Fqe_variable<ppT> out_Rx;
+    libsnark::Fqe_mul_gadget<ppT> check_out_Rx;
+    // out_Ry = theta * (G - H) - I;
+    libsnark::Fqe_variable<ppT> G_minus_H;
+    libsnark::Fqe_variable<ppT> theta_times_G_minus_H;
+    libsnark::Fqe_mul_gadget<ppT> check_theta_times_G_minus_H;
+    // out_Rz = Z1 * E;
+    libsnark::Fqe_variable<ppT> out_Rz;
+    libsnark::Fqe_mul_gadget<ppT> check_out_Rz;
+
+    bls12_377_G2_proj<ppT> out_R;
+
+    // out_coeffs.ell_0 = xi * J;
+    // out_coeffs.ell_vw = lambda;
+    // out_coeffs.ell_vv = -theta;
+    bls12_377_ate_ell_coeffs<ppT> out_coeffs;
+
+    bls12_377_ate_add_gadget(
+        libsnark::protoboard<libff::Fr<ppT>> &pb,
+        const libsnark::Fqe_variable<ppT> &base_X,
+        const libsnark::Fqe_variable<ppT> &base_Y,
```

Comment:
The base point is denoted `base` here (you pass the (`X`, `Y`) coordinates here - mixed addition), but you denote it by `Q` in the comments above, i.e. `Qy`,`Qx`. Let's add a few comment lines to either say `Qx = base_X, Qy = base_Y` or use `base_X`, `base_Y` in the comments above, otherwise this is confusing.

Reply:
used Q_X and Q_Y consistently